lib/Platform/Intl/java/com/facebook/hermes/intl/Intl.java

```diff
@@ -8,45 +8,306 @@
 package com.facebook.hermes.intl;
 
 import android.icu.util.ULocale;
+import android.os.Build;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 public class Intl {
-  // Implementer note: This method corresponds roughly to
-  // https://tc39.es/ecma402/#sec-canonicalizelocalelist
-  //
-  // Also see the implementer notes on DateTimeFormat#DateTimeFormat()
-  // for more discussion of locales and CanonicalizeLocaleList.
-  public static List<String> getCanonicalLocales(List<String> locales)
-    throws JSRangeErrorException
-  {
-    // This implementation is incomplete.  However, it does show that
-    // the Android Java ICU methods can successfully be called.
-
-    ArrayList<String> ret = new ArrayList<String>();
-    for (String l : locales) {
-      String c = ULocale.getName(l);
-      if (!ret.contains(c)) {
-        ret.add(c);
-      }
-    }
-    return ret;
-  }
-
-  // Implementer note: This method corresponds roughly to
-  // https://tc39.es/ecma402/#sup-string.prototype.tolocalelowercase
-  public static String toLocaleLowerCase(List<String> locales, String str)
-  {
-    return "lowered";
-  }
-
-  // Implementer note: This method corresponds roughly to
-  // https://tc39.es/ecma402/#sup-string.prototype.tolocaleuppercase
-  public static String toLocaleUpperCase(List<String> locales, String str)
-  {
-    return "uppered";
-  }
+
+    private static boolean isAlphaNum(String name, int min, int max) {
+        char[] chars = name.toCharArray();
```

Comment:
Could you please use `.charAt` in a simple for loop instead to avoid allocating and copying an array on each invocation? I tried it in a basic benchmark and it showed a 12x improvement.

Reply:
Hi @willholen .. Sure i'll do .. It looked so wasteful to convert the string to array, and i don't remember how i ended up with this .. Will make the changes in next iteration ..